client/consensus/grandpa/changeTree.go

```diff
@@ -0,0 +1,296 @@
+	if hash == announcingHash {
+		return false, fmt.Errorf("%w: %s", errors.New("duplicated hashes"), hash)
+	}
+
+	isDescendant, err := isDescendentOf(announcingHash, hash)
+	if err != nil {
+		return false, fmt.Errorf("cannot check ancestry: %w", err)
+	}
+
+	if !isDescendant {
+		return false, nil
+	}
+
+	if number <= pcn.change.canonHeight {
+		return false, nil
+	}
+
+	for _, childrenNodes := range pcn.children {
+		imported, err := childrenNodes.importNode(hash, number, change, isDescendentOf)
+		if err != nil {
+			return false, err
+		}
+
+		if imported {
+			return true, nil
+		}
+	}
+	childrenNode := &pendingChangeNode{
+		change: &change,
+	}
+	pcn.children = append(pcn.children, childrenNode)
+	return true, nil
+}
+
+// ChangeTree keeps track of the changes per fork allowing
+// n forks in the same structure, this structure is intended
+// to be an acyclic directed graph where the change children are
+// placed by descendency order and number, you can ensure an
+// node ancestry using the `isDescendantOfFunc`
+type ChangeTree struct {
+	roots               []*pendingChangeNode
+	bestFinalizedNumber *uint
+}
+
+// NewChangeTree create an empty ChangeTree
+func NewChangeTree() ChangeTree {
+	return ChangeTree{}
+}
+
+// Import a new node into the roots.
+//
+// The given function `is_descendent_of` should return `true` if the second
+// hash (target) is a descendent of the first hash (base).
+//
+// This method assumes that children in the same branch are imported in order.
+//
+// Returns `true` if the imported node is a root.
+// WARNING: some users of this method (i.e. consensus epoch changes roots) currently silently
+// rely on a **post-order DFS** traversal. If we are using instead a top-down traversal method
+// then the `is_descendent_of` closure, when used after a warp-sync, may end up querying the
+// backend for a block (the one corresponding to the root) that is not present and thus will
+// return a wrong result.
+func (ct *ChangeTree) Import(hash common.Hash, number uint, change PendingChange, isDescendentOf IsDescendentOf) (bool, error) {
+	for _, root := range ct.roots {
+		imported, err := root.importNode(hash, number, change, isDescendentOf)
+		if err != nil {
+			return false, err
+		}
+
+		if imported {
+			logger.Debugf("changes on header %s (%d) imported successfully",
+				hash, number)
+			return false, nil
+		}
+	}
+
+	pendingChangeNode := &pendingChangeNode{
+		change: &change,
+	}
+
+	ct.roots = append(ct.roots, pendingChangeNode)
+	return true, nil
+}
+
+// Roots returns the roots of each fork in the ChangeTree
+// This is the equivalent of the slice in the outermost layer of the roots
+func (ct *ChangeTree) Roots() []*pendingChangeNode {
+	return ct.roots
+}
+
+// GetPreOrder does a preorder traversal of the ChangeTree to get all pending changes
+func (ct *ChangeTree) GetPreOrder() []PendingChange {
+	if len(ct.roots) == 0 {
+		return nil
+	}
+
+	changes := &[]PendingChange{}
```

Comment:
I would suggest to create a slice of `PendingChange` with a capacity equals the number of elements in the roots which is the minimal number of items that can exists in the `changes` slice, then you provide the ref of `changes` to the `getPreOrder` function

```go
{
        ...
	changes := make([]PendingChange, 0, len(ct.roots))
	for i := 0; i < len(ct.roots); i++ {
		getPreOrder(&changes, ct.roots[i])
	}
	
	return changes
}

func getPreOrder(changes *[]PendingChange, changeNode *pendingChangeNode) {
	if changeNode == nil {
		return
	}

        *changes = append(*changes, *changeNode.change)
	for i := 0; i < len(changeNode.children); i++ {
		getPreOrder(changes, changeNode.children[i])
	}
}
```